openshift/oadp-operator
# OADP-1225 Add configs to support Google WIF Authentication in OADP Operator

[OADP-1225](https://issues.redhat.com//browse/OADP-1225)
We want to enable keyless cloud providers credentials like Google WIF.

- [x] NoSecret flag for Google WIF support
- [x] Remove unused velero reconciler

SAAnnotations for Google WIF support can be added to velero SA using `oc annotate`

Verification steps
1. `make deploy-olm` against this branch
2. install wif using wiki instructions https://github.com/kaovilai/k8s-config-connector/wiki/Using-Config-Connector-on-OpenShift, then configure ConfigConnector using official docs, annotating `openshift-adp` as the namespace to create resources in.
3. we would expect annotations in the namespace like so if 2. was followed.
    ```bash
    ❯ oc get ns openshift-adp -oyaml 
    apiVersion: v1
    kind: Namespace
    metadata:
      annotations:
        cnrm.cloud.google.com/project-id: openshift-gce-devel
      name: openshift-adp
    ```
4. create dpa
    - ```yaml
       # TODO:
       ```

#### Notes:

We are using ConfigConnector to [create, acquire, and/or delete](https://cloud.google.com/config-connector/docs/how-to/managing-deleting-resources) bucket resources which after configconnector is installed can be done via creating `storagebuckets.storage.cnrm.cloud.google.com` CustomResource

We create the following 
```
apiVersion: storage.cnrm.cloud.google.com/v1beta1
kind: StorageBucket
metadata:
  name: example
  namespace: openshift-adp
spec:
  resourceID: <bucket-name>
```

If everything goes well, the configconnector updates this object to something like
```
apiVersion: storage.cnrm.cloud.google.com/v1beta1
kind: StorageBucket
metadata:
  annotations:
    cnrm.cloud.google.com/management-conflict-prevention-policy: none
    cnrm.cloud.google.com/project-id: openshift-gce-devel
    cnrm.cloud.google.com/state-into-spec: merge
  resourceVersion: '1597993'
  name: example
  uid: 47653c93-5e96-4385-926a-a2ed2ff1fe20
  creationTimestamp: '2023-04-10T02:47:51Z'
  generation: 2
  managedFields:
    - apiVersion: storage.cnrm.cloud.google.com/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        'f:spec':
          .: {}
          'f:location': {}
          'f:resourceID': {}
      manager: Mozilla
      operation: Update
      time: '2023-04-10T02:47:51Z'
    - apiVersion: storage.cnrm.cloud.google.com/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        'f:metadata':
          'f:finalizers':
            .: {}
            'v:"cnrm.cloud.google.com/deletion-defender"': {}
            'v:"cnrm.cloud.google.com/finalizer"': {}
        'f:spec':
          'f:publicAccessPrevention': {}
          'f:storageClass': {}
          'f:uniformBucketLevelAccess': {}
      manager: cnrm-controller-manager
      operation: Update
      time: '2023-04-10T02:47:52Z'
    - apiVersion: storage.cnrm.cloud.google.com/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        'f:status':
          .: {}
          'f:conditions': {}
          'f:observedGeneration': {}
          'f:selfLink': {}
          'f:url': {}
      manager: cnrm-controller-manager
      operation: Update
      subresource: status
      time: '2023-04-10T02:47:52Z'
  namespace: openshift-adp
  finalizers:
    - cnrm.cloud.google.com/finalizer
    - cnrm.cloud.google.com/deletion-defender
spec:
  location: US
  publicAccessPrevention: inherited
  resourceID: <bucket-name>
  storageClass: STANDARD
  uniformBucketLevelAccess: true
status:
  conditions:
    - lastTransitionTime: '2023-04-10T02:47:52Z'
      message: The resource is up to date
      reason: UpToDate
      status: 'True'
      type: Ready
  observedGeneration: 2
  selfLink: 'https://www.googleapis.com/storage/v1/b/<bucket-name>'
  url: 'gs://<bucket-name>'
```
This confirms configconnector is able to acquire this resource.

Acronyms:
- **GSA**: Google Cloud Service Account
- **KSA**: Kubernetes Service Account

Configure applications to use Workload identity, step 4. where we will use existing serviceaccounts by creating following resource with intended service accounts to use to access our buckets
https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#config-connector
```
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMServiceAccount
metadata:
  name: [GSA_NAME]
spec:
  displayName: [DISPLAY_NAME]
```
Perhaps this step means we can use a different GSA (team specific) to access buckets than the one used to install configconnector (admin GSA that can impersonate other lower level GSAs).
We need a cluster admin to install ConfigConnector anyways since its a clusterwide operator.

Follow their step 5.: Double check that this serviceaccount [GSA_NAME] has the roles necessary to access the bucket (and to take native snapshots?) resource.

Follow their step 6. _Allow the Kubernetes service account to impersonate the IAM service account_ using ConfigConnector to use existing GSA with our KSA

Follow their step 7. _Annotate the Kubernetes service account with the email address of the IAM service account._

Follow their step 8. _Update your Pod spec to schedule the workloads on nodes that use Workload Identity and to use the annotated Kubernetes service account._
Note that the nodeSelector may not apply to us, since we are not a gke cluster, and installed configconnector

Follow their step 9. _Apply the updated configuration to your cluster:_

[verify setup](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#verify_the_setup)

🤞 [Use Workload Identity from your code](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#using_from_your_code)
> Existing code that authenticates using the instance metadata server (like code using the [Google Cloud client libraries](https://cloud.google.com/docs/authentication/client-libraries)) should work without modification.

TODO:
- [ ] Verify if we can make this work without creating new bucket
- [ ] Check backup work
